torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,1714 @@
+        state_dict_device: Optional[torch.device] = None,
+        clear_autocast_cache: bool = False,
+        force_input_to_fp32: bool = False,
+        verbose: bool = False,
+        cpu_offload: bool = False,
+    ):
+        init_start = time.time()
+        super().__init__()
+        self.process_group = process_group or get_process_group_cached()
+        self.rank = self.process_group.rank()
+        self.world_size = self.process_group.size()
+        self.reshard_after_forward = reshard_after_forward
+        self.mixed_precision = mixed_precision
+        self.fp32_reduce_scatter = fp32_reduce_scatter
+        self.flatten_parameters = flatten_parameters
+        self.move_params_to_cpu = move_params_to_cpu or cpu_offload
+        self.compute_dtype = compute_dtype or (torch.float16 if mixed_precision else torch.float32)
+        self.buffer_dtype = buffer_dtype or self.compute_dtype
+        self.move_grads_to_cpu = self.move_params_to_cpu if move_grads_to_cpu is None else move_grads_to_cpu
+        self.bucket_cap_mb = bucket_cap_mb
+        self.compute_device = compute_device or _get_default_cuda_device(module)
+        self.uncollected_opt_state: Dict[int, Dict] = {}
+        self.no_broadcast_optim_state = no_broadcast_optim_state
+        self.state_dict_device = state_dict_device or self.compute_device
+        self.clear_autocast_cache = clear_autocast_cache
+        self.force_input_to_fp32 = force_input_to_fp32
+        self.verbose = verbose
+
+        self.gradient_predivide_factor: float = self._get_gradient_predivide_factor(self.world_size)
+        self.gradient_postdivide_factor: float = self.world_size / self.gradient_predivide_factor
+
+        self.numel_padded_per_param: List[int] = []
+        self._tstart = time.time()
+
+        if self.fp32_reduce_scatter and not self.mixed_precision:
+            raise ValueError("fp32_reduce_scatter requires mixed_precision=True")
+        if self.move_params_to_cpu and not self.mixed_precision:
+            raise ValueError("cpu_offload requires mixed_precision=True")
+
+        # skip validation if the process group was created above
+        if process_group:
+            validate_process_group(self.compute_device, self.process_group)
+
+        # enable pytorch sync_bn just in case model contains sync_bn layers.
+        enable_pytorch_sync_bn(module)
+
+        # Only handle params which are not already sharded. This enables
+        # sharding individual layers of a Module, with an outer wrapper to
+        # shard any leftover parameters.
+        param_names = []
+        params = []
+        for param_name, param in module.named_parameters():
+            if not hasattr(param, "_is_sharded"):
+                param_names.append(param_name)
+                params.append(param)
+
+        self._has_params = len(params) > 0
+
+        # For now, it is either all flatten or none flatten. This will be extended to
+        # multiple flatten groups in my next PR.
```

Comment:
This seems important, as these parameter could belong to multiple different layers scattered in different parts of the model.  Running one all2all to get them all seems less ideal. I will create an issue to track this.

Reply:
sounds good, let us discuss more details there